64_FFT/main.cpp

```diff
@@ -0,0 +1,326 @@
+
+		// In contrast to fences, we just need one semaphore to rule all dispatches
+		m_timeline = m_device->createSemaphore(m_iteration);
+
+		return true;
+	}
+
+	// Ok this time we'll actually have a work loop (maybe just for the sake of future WASM so we don't timeout a Browser Tab with an unresponsive script)
+	bool keepRunning() override { return m_iteration < MaxIterations; }
+
+	// Finally the first actual work-loop
+	void workLoopBody() override
+	{
+		IQueue* const queue = getComputeQueue();
+
+		// Note that I'm using the sample struct with methods that have identical code which compiles as both C++ and HLSL
+		auto rng = nbl::hlsl::Xoroshiro64StarStar::construct({ m_iteration ^ 0xdeadbeefu,std::hash<string>()(_NBL_APP_NAME_) });
+
+		// Set to 64 to try a single subgroup's work
+		const auto elementCount = 2 * WorkgroupSize;
+		const uint32_t inputSize = sizeof(input_t) * elementCount;
+
+		// The allocators can do multiple allocations at once for efficiency
+		const uint32_t AllocationCount = 1;
+
+		// It comes with a certain drawback that you need to remember to initialize your "yet unallocated" offsets to the Invalid value
+		// this is to allow a set of allocations to fail, and you to re-try after doing something to free up space without repacking args.
+		auto inputOffset = m_upStreamingBuffer->invalid_value;
+
+		// We always just wait till an allocation becomes possible (during allocation previous "latched" frees get their latch conditions polled)
+		// Freeing of Streaming Buffer Allocations can and should be deferred until an associated polled event signals done (more on that later).
+		std::chrono::steady_clock::time_point waitTill(std::chrono::years(45));
+		// note that the API takes a time-point not a duration, because there are multiple waits and preemptions possible, so the durations wouldn't add up properly
+		m_upStreamingBuffer->multi_allocate(waitTill, AllocationCount, &inputOffset, &inputSize, &m_alignment);
+
+		// Generate our data in-place on the allocated staging buffer
+		{
+			auto* const inputPtr = reinterpret_cast<input_t*>(reinterpret_cast<uint8_t*>(m_upStreamingBuffer->getBufferPointer()) + inputOffset);
+			std::cout << "Begin array CPU\n";
+			for (auto j = 0; j < elementCount; j++)
+			{
+				float x = rng() / float(nbl::hlsl::numeric_limits<decltype(rng())>::max), y = rng() / float(nbl::hlsl::numeric_limits<decltype(rng())>::max);
+				std::cout << "(" << x << ", " << y << "), ";
+				const input_t generated(x, y);
+				// make sure our bitpatterns are in [0,1]^2 as a float
+				inputPtr[j] = generated;
+			}
+			std::cout << "\nEnd array CPU\n";
+			// Always remember to flush!
+			if (m_upStreamingBuffer->needsManualFlushOrInvalidate())
+			{
+				const auto bound = m_upStreamingBuffer->getBuffer()->getBoundMemory();
+				const ILogicalDevice::MappedMemoryRange range(bound.memory, bound.offset + inputOffset, inputSize);
+				m_device->flushMappedMemoryRanges(1, &range);
+			}
```

Comment:
random input is good for checking your forward FFT inverses properly.

However you should FFT some well known sequences as well, such as `1,0,0,....`

Reply:
Will add more tests then! The thing is that the FFT of sequences given by this does not seem to be the same as the FFT obtained via online FFT calcs. I'm not sure if it's because of the lack of bit reordering or what

Reply:
1. yes ofc the bitreverse ordering will be an issue
2. something is very wrong with your subgroup FFT ops, I compiled them in Godbolt and saw divisions by 0 + subgroup index instead of invocation index within subgroup being used

I think you either have NaNs in your output bufffer that you're downloading or complete garbage